### PR TITLE
Assets issuing and transferring support in TransactionGenerator

### PR DESCRIPTION
**Pull request by oskin1**

### CODE REVIEW

**Comment by oskin1** [comment]
#450

File: src/main/scala/org/ergoplatform/local/TransactionGenerator.scala [hunk 1]
```diff
@@ -81,6 +80,47 @@ class TransactionGenerator(viewHolder: ActorRef,
...
+    val amountToPay = (Random.nextInt(10) + 1) * 100000000
+    val paymentReq = PaymentRequest(randProposition, amountToPay, None, None)
+    val tokenPaymentReq = wallet.confirmedBalances().map { balances =>
+      if (balances.assetBalances.nonEmpty) {
+        val tokenToSpend = balances.assetBalances.toSeq(Random.nextInt(balances.assetBalances.size))
+        val tokenAmountToSpend = tokenToSpend._2 / 4
+        Algos.decode(tokenToSpend._1).map { id =>
+          PaymentRequest(randProposition, 0L, Some(Seq(Digest32 @@ id -> tokenAmountToSpend)), None)
+        }.toOption
+      } else { None }
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/local/TransactionGenerator.scala]: We do not use such a formatting

File: src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala [hunk 2]
```diff
@@ -57,7 +58,9 @@ object DefaultBoxSelector extends BoxSelector {
     }
 
     def balanceMet = currentBalance >= targetBalance
-    def assetsMet = targetAssets.forall { case (id, targetAmt) => currentAssets.getOrElse(id, 0L) >= targetAmt }
+    def assetsMet = targetAssets
+      .filterNot { case (id, _) => res.headOption.exists(bx => Algos.encode(bx.id) == id) }
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala]: What is this for?

File: src/main/scala/org/ergoplatform/local/TransactionGenerator.scala [hunk 3]
```diff
@@ -81,6 +80,47 @@ class TransactionGenerator(viewHolder: ActorRef,
...
+      inputs.headOption.map { firstInput =>
+        val assetId = Digest32 !@@ firstInput.id
+        val assetInfo = genNewAssetInfo
+        AssetIssueRequest(randProposition, assetId, assetInfo._1, assetInfo._2,
+          assetInfo._3, assetInfo._4)
+      }
+    }
+    tokenPaymentReq.flatMap { tpOutOpt =>
+      assetIssueReq.flatMap { aiOutOpt =>
+        val requests = Seq(tpOutOpt, aiOutOpt).foldLeft[Seq[TransactionRequest]](Seq(feeReq, paymentReq)) {
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/local/TransactionGenerator.scala]: Might be rewritten easier, e.g. `Seq(tpOutOpt, aiOutOpt, Some(feeReq), Some(paymentReq)).flatten`

File: src/main/scala/org/ergoplatform/local/TransactionGenerator.scala [hunk 4]
```diff
@@ -81,6 +80,47 @@ class TransactionGenerator(viewHolder: ActorRef,
 
     case SuccessfulTransaction(_) => self ! Attempt
   }
+
+  private def genTransaction(wallet: ErgoWallet): Future[Try[ErgoTransaction]] = {
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/local/TransactionGenerator.scala]: The idea of transaction generator is to simulate user activity. Thus `genTransaction` should generate transactions, similar to regular users. I doubt regular users will create transactions, that send Ergo, assets and issue new assets at the same time. Let's rewrite it in such a way, that in most cases it generate regular transaction that send ergo, in few cases it generate transactions that send some tokens, in rare cases it generate transactions that issue a new token.

File: src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala [hunk 5]
```diff
@@ -75,20 +78,18 @@ object DefaultBoxSelector extends BoxSelector {
...
       //If this condition is satisfied on the previous step, we will do one extra check (which is not that much).
-      if (pickBoxes(inputBoxes, bc => externalFilter(bc) && bc.assets.exists { case (id, _) =>
+      if (pickBoxes(inputBoxes, box => externalFilter(box) && box.assets.exists { case (id, _) =>
         val targetAmt = targetAssets.getOrElse(id, 0L)
         lazy val currentAmt = currentAssets.getOrElse(id, 0L)
         targetAmt > 0 && targetAmt > currentAmt
       }, assetsMet)) {
-        subtractAssets(currentAssets, targetAssets)
+        val notNewTargetedAssets =
+          targetAssets.filterNot { case (id, _) => res.headOption.exists(bx => Algos.encode(bx.id) == id) }
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala]: why head box is dedicated here?

File: src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala [hunk 6]
```diff
@@ -75,20 +78,18 @@ object DefaultBoxSelector extends BoxSelector {
...
         val changeBoxesAssets: Seq[mutable.Map[ModifierId, Long]] = currentAssets.grouped(ErgoBox.MaxTokens).toSeq
         val changeBalance = currentBalance - targetBalance
         formChangeBoxes(changeBalance, changeBoxesAssets).map(changeBoxes => BoxSelectionResult(res, changeBoxes))
-      } else {
-        None
-      }
-    } else {
-      None
-    }
+      } else { None }
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala]: previous formatting was correct

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala [hunk 7]
```diff
@@ -144,8 +144,8 @@ class ErgoWalletActor(settings: ErgoSettings) extends Actor with ScorexLogging {
 
   protected def generateTransactionWithOutputs(payTo: Seq[ErgoBoxCandidate]): Try[ErgoTransaction] = Try {
     require(prover.dlogPubkeys.nonEmpty, "No public keys in the prover to extract change address from")
-    require(payTo.forall(_.value > 0), "Non-positive Ergo value")
-    require(payTo.forall(_.additionalTokens.forall(_._2 > 0)), "Non-positive asset value")
+    require(payTo.forall(_.value >= 0), "Non-positive Ergo value")
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala]: Why is this related to current PR?

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala [hunk 8]
```diff
@@ -42,13 +40,15 @@ class ErgoWalletActor(settings: ErgoSettings) extends Actor with ScorexLogging {
...
 
-  private implicit val addressEncoder = ErgoAddressEncoder(settings)
+  private implicit val addressEncoder: ErgoAddressEncoder = ErgoAddressEncoder(settings)
   private val publicKeys: Seq[P2PKAddress] = Seq(prover.dlogPubkeys: _ *).map(P2PKAddress.apply)
 
   private val trackedAddresses: mutable.Buffer[ErgoAddress] = publicKeys.toBuffer
 
   private val trackedBytes: mutable.Buffer[Array[Byte]] = trackedAddresses.map(_.contentBytes)
 
+  private def filterFn(trackedBox: TrackedBox) = trackedBox.chainStatus.onchain
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala]: If you moved the function, please also move comments related to it.

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala [hunk 9]
```diff
@@ -198,6 +197,9 @@ class ErgoWalletActor(settings: ErgoSettings) extends Actor with ScorexLogging {
 
     case ReadTrackedAddresses =>
       sender() ! trackedAddresses.toIndexedSeq
+
+    case ReadInputsFor(outputs) =>
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala]: I don't like that transaction generator requires new functionality in wallet. It is possible to transfer Ergo, transfer assets and issue assets without `ReadInputsFor` request, correct? Why do you need it to generate transactions?

File: src/test/scala/org/ergoplatform/mining/ErgoMinerSpec.scala [hunk 10]
```diff
@@ -99,7 +99,9 @@ class ErgoMinerSpec extends FlatSpec with ErgoTestHelpers with ValidBlocksGenera
 
         val unsignedTx = new UnsignedErgoTransaction(inputs, outputs)
         val tx = prover.sign(unsignedTx, IndexedSeq(boxToSend), ErgoStateContext(r.h.fullBlockHeight, r.s.rootHash)).get
-        nodeViewHolderRef ! LocallyGeneratedTransaction[ErgoTransaction](tx)
```
Comment (catena2w) [inline src/test/scala/org/ergoplatform/mining/ErgoMinerSpec.scala]: why?

File: src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala (hunk 2 above)
Comment (oskin1) [reply src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala]: @catena2w, Box selector considered outputs with issuing assets when calculating targeted asset balance, this led to selection failure. This is for excluding new assets.

File: src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala (hunk 5 above)
Comment (oskin1) [reply src/main/scala/org/ergoplatform/nodeView/wallet/DefaultBoxSelector.scala]: @catena2w, this is how we can define wether new assets are presented in the sequence  or not.

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala (hunk 7 above)
Comment (oskin1) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala]: Because old requirement did not allow to create asset issuing outputs.

File: src/test/scala/org/ergoplatform/mining/ErgoMinerSpec.scala (hunk 10 above)
Comment (oskin1) [reply src/test/scala/org/ergoplatform/mining/ErgoMinerSpec.scala]: @catena2w, Old approach led to rediculously long logs in Travis so it aborts tests exucution.

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala (hunk 9 above)
Comment (oskin1) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala]: @catena2w, to issue new asset we need to know the first input for transaction in advance - so this functionality is required.

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala [hunk 11]
```diff
@@ -142,44 +146,75 @@ class ErgoWalletActor(settings: ErgoSettings) extends Actor with ScorexLogging {
...
+      case other => throw new Exception(s"Unknown TransactionRequest type: $other")
+    }
+  }
 
-    val targetBalance = payTo.map(_.value).sum
+  protected def generateTransactionWithOutputs(requests: Seq[TransactionRequest]): Try[ErgoTransaction] =
+    requestsToBoxCandidates(requests).map { payTo =>
+      require(prover.dlogPubkeys.nonEmpty, "No public keys in the prover to extract change address from")
+      require(payTo.forall(_.value >= 0), "Non-positive Ergo value")
+      require(payTo.forall(_.additionalTokens.forall(_._2 >= 0)), "Non-positive asset value")
```
Comment (catena2w) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala]: I think we should also require at most 1 AssetIssueRequest

Comment (oskin1) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala]: done
